File: Parser/Impl/LL.cpp
#include "Parser/Impl/LL.hpp"

#include <algorithm>

namespace Parser::Impl
{
    LLBase::LLBase(const Grammar &grammar)
    : Base(grammar)
    {
        std::vector<std::set<unsigned int>> firstSets;
        std::vector<std::set<unsigned int>> followSets;
        std::set<unsigned int> nullableNonterminals;
        mGrammar.computeSets(firstSets, followSets, nullableNonterminals);

        mValid = computeParseTable(firstSets, followSets, nullableNonterminals);
    }

    bool LLBase::addParseTableEntry(unsigned int rule, unsigned int symbol, unsigned int rhs)
    {
        if(mParseTable.at(rule, symbol) == UINT_MAX) {
            mParseTable.at(rule, symbol) = rhs;
            return true;
        } else {
            mConflict.rule = rule;
            mConflict.symbol = symbol;
            mConflict.rhs1 = mParseTable.at(rule, symbol);
            mConflict.rhs2 = rhs;
            return false;
        }
    }

    bool LLBase::addParseTableEntries(unsigned int rule, const std::set<unsigned int> &symbols, unsigned int rhs)
    {
        for(unsigned int s : symbols) {
            if(!addParseTableEntry(rule, s, rhs)) {
                return false;
            }
        }

        return true;
    }

    bool LLBase::computeParseTable(const std::vector<std::set<unsigned int>> &firstSets, std::vector<std::set<unsigned int>> &followSets, std::set<unsigned int> &nullableNonterminals)
    {
        mParseTable.resize(mGrammar.rules().size(), mGrammar.terminals().size(), UINT_MAX);

        for(unsigned int i=0; i<mGrammar.rules().size(); i++) {
            const Grammar::Rule &rule = mGrammar.rules()[i];

            for(unsigned int j=0; j<rule.rhs.size(); j++) {
                const Grammar::Symbol &symbol = rule.rhs[j][0];
                switch(symbol.type) {
                    case Grammar::Symbol::Type::Terminal:
                        if(!addParseTableEntry(i, symbol.index, j)) {
                            return false;
                        }
                        break;
                    
                    case Grammar::Symbol::Type::Nonterminal:
                        if(!addParseTableEntries(i, firstSets[symbol.index], j)) {
                            return false;
                        }

                        if(nullableNonterminals.contains(symbol.index)) {
                            if(!addParseTableEntries(i, followSets[symbol.index], j)) {
                                return false;
                            }
                        }
                        break;
                    
                    case Grammar::Symbol::Type::Epsilon:
                        if(!addParseTableEntries(i, followSets[i], j)) {
                            return false;
                        }
                        break;
                }
            }
        }

        return true;
    }

    bool LLBase::valid() const
    {
        return mValid;
    }

    const LLBase::Conflict &LLBase::conflict() const
    {
        return mConflict;
    }

    unsigned int LLBase::rhs(unsigned int rule, unsigned int symbol) const
    {
        if(symbol == Tokenizer::kErrorTokenValue) {
            return UINT_MAX;
        } else {
            return mParseTable.at(rule, symbol);
        }
    }

    bool LLBase::runParse(Tokenizer::Stream &stream, ParseStackBase &parseStack) const
    {
        struct PredictItem {
            enum class Type {
                Terminal,
                Nonterminal,
                Reduce
            };
            Type type;
            union {
                struct {
                    unsigned int index;
                    unsigned int rule;
                    unsigned int pos;
                } symbol;
                struct {
                    unsigned int rule;
                    unsigned int parseStackStart;
                } reduce;
            };
        };

        std::vector<PredictItem> predictStack;

        predictStack.push_back(PredictItem{PredictItem::Type::Nonterminal, grammar().startRule()});

        while(predictStack.size() > 0) {
            PredictItem predictItem = predictStack.back();
            predictStack.pop_back();

            switch(predictItem.type) {
                case PredictItem::Type::Terminal:
                {
                    if(stream.nextToken().value == predictItem.symbol.index) {
                        shift(stream.nextToken(), parseStack);
                        
                        auto it2 = mMatchListeners.find(predictItem.symbol.rule);
                        if(it2 != mMatchListeners.end()) {
                            it2->second(predictItem.symbol.pos);
                        }
                        stream.consumeToken();
                    } else {
                        return false;
                    }
                    break;
                }

                case PredictItem::Type::Nonterminal:
                {
                    unsigned int nextRule = predictItem.symbol.index;
                    unsigned int nextRhs = rhs(nextRule, stream.nextToken().value);

                    if(nextRhs == UINT_MAX) {
                        return false;
                    }   

                    if(canReduce(nextRule)) {
                        predictStack.push_back(PredictItem{PredictItem::Type::Reduce, nextRule, (unsigned int)parseStack.size()});
                    }

                    const std::vector<Grammar::Symbol> &symbols = grammar().rules()[nextRule].rhs[nextRhs];
                    for(unsigned int i=0; i<symbols.size(); i++) {
                        unsigned int ri = (unsigned int)symbols.size() - i - 1;
                        const Grammar::Symbol &s = symbols[ri];
                        switch(s.type) {
                            case Grammar::Symbol::Type::Terminal:
                                predictStack.push_back(PredictItem{PredictItem::Type::Terminal, s.index, nextRule, ri});
                                break;
                            
                            case Grammar::Symbol::Type::Nonterminal:
                                predictStack.push_back(PredictItem{PredictItem::Type::Nonterminal, s.index, nextRule, ri});
                                break;
                            
                            case Grammar::Symbol::Type::Epsilon:
                                break;
                        }
                    }
                    break;
                }

                case PredictItem::Type::Reduce:
                {
                    unsigned int currentRule = predictItem.reduce.rule;
                    unsigned int parseStackStart = predictItem.reduce.parseStackStart;

                    reduce(currentRule, parseStackStart, parseStack);
                    break;
                }
            }
        }
    
        return true;
    }
}

File: src/parser/algorithm/ll.rs
mod parse_table;
use parse_table::{Conflict, ParseTable};

use crate::parser;
use parser::Grammar;
use parser::tokenizer::Stream;

use std::collections::HashMap;

type ParseItem<T> = (usize, Option<T>);
type TerminalDecorator<T> = dyn Fn(&parser::tokenizer::Token) -> T;
type Reducer<T> = dyn Fn(&[ParseItem<T>]) -> T;

pub struct LL<T> {
    pub grammar: Grammar,
    parse_table: ParseTable,
    terminal_decorators: HashMap<usize, Box<TerminalDecorator<T>>>,
    reducers: HashMap<usize, Box<Reducer<T>>>
}

type ParseStack<T> = Vec<ParseItem<T>>;

enum PredictItem {
    Terminal(usize),
    Nonterminal(usize),
    Reduce(usize, usize)
}

#[allow(dead_code)]
impl<T> LL<T> {
    pub fn new(grammar: Grammar) -> Result<LL<T>, Conflict> {
        let parse_table= ParseTable::new(&grammar)?;
        let ll = Self { grammar, parse_table, terminal_decorators: HashMap::new(), reducers: HashMap::new() };
        return Ok(ll);
    }

    pub fn add_terminal_decorator<F>(&mut self, terminal: &str, terminal_decorator: F)
    where F: 'static + Fn(&parser::tokenizer::Token) -> T {
        if let Some(idx) = self.grammar.terminal_index(terminal) {
            self.terminal_decorators.insert(idx, Box::new(terminal_decorator));
        }
    }

    pub fn add_reducer<F>(&mut self, rule: &str, reducer: F) 
    where F: 'static + Fn(&[ParseItem<T>]) -> T {
        if let Some(idx) = self.grammar.rule_index(rule) {        
            self.reducers.insert(idx, Box::new(reducer));
        }
    }

    pub fn parse(&self, mut stream: Stream) -> Option<T> {
        let mut parse_stack: ParseStack<T> = ParseStack::new();
        let mut predict_stack = Vec::new();

        predict_stack.push(PredictItem::Nonterminal(self.grammar.start_rule));

        while let Some(predict_item) = predict_stack.pop() {
            match predict_item {
                PredictItem::Terminal(terminal) => {
                    if stream.next_token().value == terminal {
                        self.shift(stream.next_token(), &mut parse_stack);

                        stream.consume_token();
                    } else {
                        return None;
                    }
                },
                PredictItem::Nonterminal(rule) => {
                    let next_rule = rule;
                    let next_rhs = self.parse_table.rhs(rule, stream.next_token().value);

                    if next_rhs == usize::MAX {
                        return None;
                    }

                    if self.can_reduce(next_rule) {
                        predict_stack.push(PredictItem::Reduce(next_rule, parse_stack.len()));
                    }

                    let symbols = &self.grammar.rules[next_rule].rhs[next_rhs];
                    for i in 0..symbols.len() {
                        let ri = symbols.len() - 1 - i;
                        let s = &symbols[ri];
                        match s {
                            parser::grammar::Symbol::Terminal(terminal_index) => predict_stack.push(PredictItem::Terminal(*terminal_index)),
                            parser::grammar::Symbol::Nonterminal(rule_index) => predict_stack.push(PredictItem::Nonterminal(*rule_index)),
                            parser::grammar::Symbol::Epsilon => ()
                        }
                    }
                },
                PredictItem::Reduce(current_rule, parse_stack_start) => {
                    self.reduce(current_rule, parse_stack_start, &mut parse_stack);
                }
            }
        }

        if let Some((_, Some(data))) = parse_stack.pop() {
            return Some(data);
        } else {
            return None;
        }
    }

    fn shift(&self, token: &parser::tokenizer::Token, parse_stack: &mut ParseStack<T>) {
        let data;
        if let Some(terminal_decorator) = self.terminal_decorators.get(&token.value) {
            data = Some(terminal_decorator(token));
        } else {
            data = None;
        }

        let parse_item = (token.value, data);
        parse_stack.push(parse_item);
    }

    fn can_reduce(&self, rule: usize) -> bool {
        return self.reducers.contains_key(&rule);
    }

    fn reduce(&self, rule: usize, stack_start: usize, parse_stack: &mut ParseStack<T>) {
        if let Some(reducer) = self.reducers.get(&rule) {
            let data = reducer(&parse_stack[stack_start..]);
            parse_stack.drain(stack_start..);

            let parse_item = (rule, Some(data));
            parse_stack.push(parse_item);
        }
    }
}
